### ISSUE TITLE
Sampling-path corruption: single garbage token injected into ~11% of completions when temperature > 0 (Metal backend, Apple Silicon)

### ISSUE THREAD

**Issue by swifty99** [vllm-project/vllm-metal#622]
### Summary

With any non-zero temperature, `vllm serve` on the Metal backend injects a
single stray token into roughly **1 in 9 completions** (11.1%, 5/45 in our
largest run; small-n, so treat the exact percentage as an order-of-magnitude
estimate, not a precise rate — see Evidence for the full run and two smaller
independent runs landing in the same 10–20% band). The
substitution can land in a tool-call argument value, in a tool-call
**function name** (producing a call to a function that does not exist), or
in free-text reasoning output. Greedy decoding (`temperature=0`) is clean
over 90 samples total across several separate runs on the same host.

This looks like a decode-path bug, not a tool-call-parser bug: it
reproduces before any parser sees the output (confirmed against the raw
`content` field with no `--tool-call-parser` configured at all), it hits
plain reasoning text as well as structured arguments, and it disappears
entirely at `temperature=0`.

**Impact:** any application relying on non-greedy sampling (i.e. following
this model family's own recommended operating point:
`temperature=0.7, top_p=0.8, top_k=20, repetition_penalty=1.05`) gets
silently corrupted tool calls at production scale. We only caught it
because our scoring harness checks tool-call arguments field-by-field
against an expected value; a fuzzier judge (or no judge at all) would have
let this through as "looks plausible."

### Environment

- `vllm`: `0.27.1+cpu` (reported by `/version` and `vllm --version`)
- `vllm-metal`: `0.3.0.dev20260816085229`
- Platform plugin: `metal` (`vllm.platform_plugins` -> `metal -> vllm_metal:register`)
- OS: macOS 26.5.1 (BuildVersion 25F80), arm64
- Hardware: Apple M3 Ultra, 96 GB unified memory
- Model: `mlx-community/Qwen3.8-27B-8bit` (MLX-quantized, served directly by
  vllm-metal — no conversion step)
- Server flags: `--enable-auto-tool-choice --tool-call-parser qwen3_xml
  --reasoning-parser qwen3 --max-num-seqs 1`
- No `--api-key` set (any bearer token accepted)

Full server command used for every run below:

```bash
vllm serve mlx-community/Qwen3.8-27B-8bit \
  --served-model-name qwen3.8-27b-8bit \
  --host 0.0.0.0 --port 8000 \
  --enable-auto-tool-choice \
  --tool-call-parser qwen3_xml \
  --reasoning-parser qwen3 \
  --max-num-seqs 1
```

### Cross-check that rules out model / quantization / harness

Same model weights, same prompts, same harness, two other inference
engines, zero corruption:

| Engine | Quantization | temperature=0.7, n=45 |
|---|---|---|
| **vllm-metal** | MLX 8bit | **40/45 clean (89%)** |
| rapid-mlx (MLX's own server) | MLX 8bit | 45/45 clean |
| llama.cpp (`llama-server`, `--jinja`) | GGUF Q8_0 | 45/45 clean |

Full numbers: `BEFUND-engine-sampling-campaign-0817.md` in the repo this
issue was drafted from.

### Reproduction (for an agent or script, not just a human)

Two parts: start the server, then run a scripted probe that fires N
requests and reports the corruption rate. Expect **~10–20% at
temperature=0.7**, **0% at temperature=0**.

**1. Install and start the server** (Apple Silicon host, call it `<HOST>`):

```bash
curl -fsSL https://raw.githubusercontent.com/vllm-project/vllm-metal/main/install.sh | bash
huggingface-cli download mlx-community/Qwen3.8-27B-8bit   # ~28 GB

~/.venv-vllm-metal/bin/vllm serve mlx-community/Qwen3.8-27B-8bit \
  --served-model-name qwen3.8-27b-8bit \
  --host 0.0.0.0 --port 8000 \
  --enable-auto-tool-choice \
  --tool-call-parser qwen3_xml \
  --reasoning-parser qwen3 \
  --max-num-seqs 1
# wait for "Application startup complete" / GET /v1/models to return 200
```

**2. Run this probe** (stdlib only, no dependencies — save as `probe.py` and
run `python3 probe.py http://<HOST>:8000 50`):

```python
#!/usr/bin/env python3
"""Fires N chat-completion requests with a tool schema at temperature 0.7,
flags any completion containing a non-ASCII stray token in a tool-call
argument, tool-call function name, or the reasoning/content text. Expects
a corruption rate near 0% at --sampling greedy (edit TEMPERATURE below to
0.0 for the control run) and roughly 10-20% at temperature 0.7."""
import json, sys, urllib.request

ENDPOINT = sys.argv[1] if len(sys.argv) > 1 else "http://localhost:8000"
N = int(sys.argv[2]) if len(sys.argv) > 2 else 50
TEMPERATURE = 0.7   # set to 0.0 for the greedy control run

PAYLOAD = {
    "model": "qwen3.8-27b-8bit",
    "messages": [
        {"role": "system", "content": "Du bist Hermes. Nutze die bereitgestellten Tools."},
        {"role": "user", "content": "Zeig mir die letzten 50 Logzeilen des Containers hermes."},
    ],
    "tools": [{
        "type": "function",
        "function": {
            "name": "docker_logs",
            "description": "Liest Logzeilen eines Docker-Containers.",
            "parameters": {
                "type": "object",
                "properties": {
                    "container": {"type": "string"},
                    "lines": {"type": "integer", "description": "Anzahl der Zeilen"},
                },
                "required": ["container", "lines"],
            },
        },
    }],
    "tool_choice": "auto",
    "temperature": TEMPERATURE,
    "top_p": 0.8,
    "top_k": 20,
    "repetition_penalty": 1.05,
    "max_tokens": 512,
}

VALID_FN = {"docker_logs"}

def has_stray_token(s: str) -> bool:
    # crude but effective: the injected tokens observed so far are CJK/Thai/
    # box-drawing characters that have no business in an English/German
    # numeric-argument completion (e.g. U+65E2 "既", U+7684 "的", U+0E1E "ฯ").
    return any(ord(c) > 0x2E80 for c in s)

bad = 0
for i in range(N):
    req = urllib.request.Request(
        f"{ENDPOINT}/v1/chat/completions",
        data=json.dumps(PAYLOAD).encode(),
        headers={"Content-Type": "application/json", "Authorization": "Bearer dummy"},
    )
    with urllib.request.urlopen(req, timeout=60) as resp:
        msg = json.load(resp)["choices"][0]["message"]
    tool_calls = msg.get("tool_calls") or []
    reasoning = msg.get("reasoning") or msg.get("content") or ""
    corrupt = has_stray_token(reasoning)
    for tc in tool_calls:
        fn = tc["function"]["name"]
        args = tc["function"]["arguments"]
        if fn not in VALID_FN or has_stray_token(args):
            corrupt = True
    marker = "CORRUPT" if corrupt else "ok"
    if corrupt:
        bad += 1
        print(f"[{i+1}/{N}] {marker}  tool_calls={tool_calls!r}")
    else:
        print(f"[{i+1}/{N}] {marker}")

print(f"\n{bad}/{N} corrupted ({100*bad/N:.1f}%) at temperature={TEMPERATURE}")
```

### Expected vs. actual

- **Expected:** at `temperature=0.7` (this model family's documented
  recommended operating point), 0/N completions contain stray tokens —
  same as at `temperature=0`.
- **Actual:** ~10–20% of completions (11.1% at n=45, our largest run so
  far) contain exactly one injected token, either in a tool-call argument
  value, a tool-call function name, or free-text reasoning.

### Evidence: three raw completions from our n=45 run

Server started exactly as in "Environment" above, request sampling
`{temperature: 0.7, top_p: 0.8, top_k: 20, repetition_penalty: 1.05}`.

**1. Argument value corrupted** (prompt: "Zeig mir die letzten 50
Logzeilen des Containers hermes.", tool `docker_logs(container, lines:
integer)`):

```json
{
  "tool_calls": [{
    "function": {
      "name": "docker_logs",
      "arguments": "{\"container\": \"hermes\", \"lines\": \"50\\n</既定\"}"
    }
  }],
  "reasoning": "The user is asking for the last 50 log lines of the container \"hermes\". I should use the docker_logs tool with container=\"hermes\" and lines=50.\n"
}
```

Note `lines` was supposed to be the integer `50`; instead it's the string
`"50\n</既定"` — `既定` (U+65E2 U+5B9A) has no plausible source in this
context (German/English prompt, no CJK anywhere in the conversation).
`reasoning` here is clean, so the corruption is localized to the tool-call
argument serialization, not a general text-quality problem in this sample.

**2. Reasoning text ALSO corrupted, not just structured output** (prompt:
"Suche alle Rechnungen vom 1. Juli 2026.", tool `search_invoices(date_iso:
string)`):

```json
{
  "tool_calls": [{
    "function": {
      "name": "search_invoices",
      "arguments": "{\"date_iso\": \"2026-的共同07-01\"}"
    }
  }],
  "reasoning": "The user wants to search for kolesterol invoices from July 1, 2026. Let's use the search_invoices tool with the date 2026-07-01.\n"
}
```

Two independent injections in the same completion: `的共同` (U+7684 U+5171
U+540C) inside the date argument, and the nonsense word "kolesterol"
spliced into otherwise-fluent English reasoning ("search for kolesterol
invoices" — not a real English word, no relation to the domain). This is
why we don't think this is specific to tool-call serialization: it also
corrupts plain generated prose in the same response.

**3. Function name itself corrupted, not just an argument** (prompt: "Wie
ist der aktuelle Zustand von light.kueche?", tools `hactl_get_state(entity_id)`
/ `hactl_call_service(service, entity_id)`):

```json
{
  "tool_calls": [{
    "function": {
      "name": "hactl_getcoles",
      "arguments": "{\"entity_id\": \"light.kueche\"}"
    }
  }],
  "reasoning": "The user is asking about the current state of light.kueche (kitchen light). I should use the hactl_get_state tool to read the state of that entity.\n"
}
```

`reasoning` correctly names `hactl_get_state` (the only tool that matches
the description). The emitted `tool_calls[0].function.name` is
`hactl_getcoles` — not `hactl_get_state`, not any other tool in the
request's `tools` array, not a hallucinated-but-plausible name. This is a
call to a function that does not exist; a client would get a hard failure,
not a wrong-but-parseable answer. We read this as the clearest sign that
the substitution happens at the token-sampling level, after the correct
plan was already formed (visible in `reasoning`), rather than being a
higher-level model mistake.

### Additional notes

- **Not a tool-call-parser artifact.** An earlier run with no
  `--tool-call-parser` configured at all showed the same single-token
  injection sitting in the raw `content` field, before any parsing:
  `<parameter=lines>\n50\n</既定\n</parameter>` instead of
  `<parameter=lines>\n50\n</parameter>`. Parser and client-side harness are
  both ruled out.
- **Not top_p/top_k/repetition_penalty specifically.** Isolated each
  parameter earlier (9 samples each): `temperature=0.7` alone already
  corrupts at ~1/9; adding `top_p`, `top_k`, or `repetition_penalty`
  individually stays in the same 1–2/9 range. It's `temperature > 0` in
  general, not one particular sampling knob.
- **Not (obviously) deterministic per-request.** Two early runs on
  identical requests came back byte-identical including the corruption,
  which briefly looked like a deterministic indexing bug — that did not
  hold up: it's the expected outcome of a low-probability event on short,
  strongly-peaked completions, not a fixed seed artifact.
- We're happy to run further probes against this host if useful — the
  model and harness that produced this report are still in place.

**Comment by evemcgivern** [comment]
Independent reproduction on different hardware, with three data points that may help narrow it.

**Environment** — M4 Max (128 GB), macOS 26.5.2 arm64, `vllm-metal 0.3.0.dev20260817081527` (one day newer than the build in the report), `vllm 0.27.1`, same model `mlx-community/Qwen3.8-27B-8bit`, same `--enable-auto-tool-choice --tool-call-parser qwen3_xml`.

**1. It reproduces at temperature 0.2, not just at the model's recommended operating point.** We were running greedy-ish sampling precisely to keep tool calls deterministic and still saw it. That extends the affected range well below the `temperature=0.7` case in the report — consistent with "any non-zero temperature."

**2. Prefix caching is not implicated.** We measured it both ways over three full agent runs each:

| Prefix caching | Transcript chars | Corrupted tokens | Rate |
|---|---|---|---|
| `--enable-prefix-caching` | 86,749 | 178 | 2,052 / M chars |
| `--no-enable-prefix-caching` | 119,613 | 365 | 3,051 / M chars |

Corruption is present in **all six runs** and is, if anything, *higher* with caching disabled (that arm simply generated more text). We had initially suspected `mamba_cache_mode=align`; these numbers ruled it out and pointed us at the sampling path, which is what led us here.

**3. A clean control on identical weights.** The same model, same quantization, same fixture and same prompts served by **vllm-mlx** instead of vllm-metal produced **zero** corrupted tokens across three runs. That isolates it to this engine's decode path rather than the weights, the quantization, or the chat template.

**Token profile** — scattered single characters from other scripts, one per occurrence: Cyrillic (`а е о в с м ч у т и р п й л`), Arabic (`ا ر`), CJK (`的 既定 星光 저작권자`), plus `é` and `≈`. One landed inside a function name exactly as described — `<function既定>` — producing a call to a nonexistent function and stalling our agent loop at that iteration.

**One caveat on comparing rates:** ours are corrupted *tokens per character of transcript* across long multi-turn agent runs, not the report's per-completion rate, so they aren't directly comparable to the 11% figure — they're a different denominator, offered as corroboration of presence and of the cache-independence, not as a refinement of the percentage.

We're switching to `temperature=0` on the strength of this issue. Happy to run further characterization on M4 Max if it would help — e.g. a temperature sweep to find where the rate becomes non-zero.

**Comment by LxYuan0420** [comment]
Thanks, the repro is useful. Can you try a smaller no-tool repro with one plain prompt, fixed seed, raw token IDs, and logprobs/top-k around the first bad token? Tool use makes the issue easier to notice, but it also adds parser/schema complexity, so a plain prompt will tell us whether this is really in the logits/sampling path.

Also try the same prompt with `top_k=1`. If the sampled token is outside the candidate set, we can trace the logits bridge and sampler path.

**Comment by evemcgivern** [comment]
Ran the smaller no-tool repro you asked for. It reproduces on a plain `/v1/completions` prompt with no tools and no parser involved, and the `top_k=1` control is informative.

**Setup:** `mlx-community/Qwen3.8-27B-8bit`, `vllm-metal 0.3.0.dev20260817081527`, `vllm 0.27.1`, macOS 26.5.2 arm64, M4 Max. Plain prose prompt, `temperature=0.7, top_p=0.8, logprobs=20`, seeds swept from 0. Seed 2 reproduces.

## First bad token — seed 2, position 43

Emitted: `' рег'`, **logprob -29.1435** (p ≈ 2e-13)

Context: `matches. This keeps the map рег working without needing`

Top-20 candidates at that position:

```
    -0.5498  ' fast'          -5.9248  ' accurate'
    -1.7998  ' working'       -6.1748  ' usable'
    -2.0498  ' simple'        -6.2998  ' running'
    -2.5498  ' quick'         -6.9248  ' neat'
    -4.1748  ' from'          -7.0498  ' mostly'
    -4.9248  ' organized'     -7.1748  ' work'
    -5.2998  ' useful'        -7.5498  ' moving'
    -5.5498  ' easy'          -7.5498  ' speedy'
    -5.7998  ' tidy'          -7.6748  ' small'
                              -8.0498  ' orderly'
                              -8.1748  ' correct'
   -29.1435  ' рег'   <-- EMITTED (21st entry)
```

Note the emitted token is the **21st** entry in a `logprobs=20` response — it is present only because vLLM appends the sampled token to `top_logprobs`. It is **not** among the genuine top-20: every real candidate sits at ≥ -8.17, and the emitted one is 21 nats below the argmax.

## Controls, same prompt and seed

| Run | Bad positions |
|---|---|
| `temperature=0.7, top_p=0.8` | 43, 67, 76 |
| **`top_k=1`** (same temperature) | **43, 70, 76, 210** |
| `temperature=0` | **none — clean** |

**`top_k=1` does not prevent it.** With `top_k=1` only the argmax is eligible, so a -29.1 token being emitted puts the fault **downstream of top-k filtering** — the selected index does not correspond to the filtered candidate set. Greedy remains clean.

That seems to rule out the tool-call parser, the chat template, and candidate construction, and points at the sampling/selection step in the non-greedy path.

## One correction to the issue's framing

We measured greedy over a much larger sample than 90 completions and it is **not completely clean** — across 78,421 characters of agent transcripts at `temperature=0` we saw 8 corrupted tokens (~102 per million), versus ~2,052 per million at `temperature=0.2`. So greedy is roughly a **95% reduction rather than an elimination**. Worth knowing if the fix is validated against greedy output.

Repro script is short and self-contained (plain urllib + `/v1/completions`, sweeps seeds, dumps logprobs around the first bad token, then runs both controls) — happy to attach it or open it as a test if useful.

**Comment by ricky-chaoju** [comment]
I traced this down. The non-greedy path runs vLLM's torch sampler on MPS, and torch's `exponential_()` on MPS sometimes returns exact zeros (873 in 29 billion draws in my test; CPU gives none). The sampler computes `probs.div(q).argmax()`, so a zero draw becomes inf or NaN and argmax picks a random vocab id, which on Qwen's vocab shows up as a random CJK character. Clamping just the zeros in a harness removed every bad token, so this is the whole mechanism.

@evemcgivern this fits your `top_k=1` result: masked positions are `0/0 = NaN` and argmax still picks them. Were your `temperature=0` runs requesting logprobs? Greedy with logprobs also goes through the torch sampler, so that residual would be the same bug.

#629 pins torch sampling to CPU. Reproduced 6/60 corrupted completions before the fix with the original request shape on Qwen3.8-27B-4bit, 0/60 after.

**Comment by ricky-chaoju** [comment]
I traced this down. The non-greedy path runs vLLM's torch sampler on MPS, and torch's `exponential_()` on MPS sometimes returns exact zeros (873 in 29 billion draws in my test; CPU gives none). The sampler computes `probs.div(q).argmax()`, so a zero draw becomes inf or NaN and argmax picks a random vocab id, which on Qwen's vocab shows up as a random CJK character. Clamping just the zeros in a harness removed every bad token, so this is the whole mechanism.

@evemcgivern this fits your `top_k=1` result: masked positions are `0/0 = NaN` and argmax still picks them. Were your `temperature=0` runs requesting logprobs? Greedy with logprobs also goes through the torch sampler, so that residual would be the same bug.

#629 pins torch sampling to CPU. Reproduced 6/60 corrupted completions before the fix with the original request shape on Qwen3.8-27B-4bit, 0/60 after.
